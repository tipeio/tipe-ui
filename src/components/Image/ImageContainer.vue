<template>
  <error-template v-if="status === 'error'"/>
  <tipe-image v-else-if="status === 'ok'"/>
  <waiting-template v-else/>
</template>

<script>
import vueTypes from 'vue-types'
import { Machine } from 'xstate'
import imageLoader from '../../libs/imageLoader'

import TipeImage from './Image.vue'
import ErrorTemplate from './templates/ErrorTemplate.vue'
import WaitingTemplate from './templates/WaitingTemplate.vue'

const status = Machine({
  initial: 'waiting',
  states: {
    waiting: {
      on: {
        done: 'ok',
        error: 'error'
      }
    },
    ok: {},
    error: {}
  }
})

export default {
  name: 'TipeImageContainer',
  components: {
    TipeImage,
    ErrorTemplate,
    WaitingTemplate
  },
  props: {
    alt: vueTypes.string.def(''),
    url: vueTypes.string.def(''),
    __imageLoader__: vueTypes.func.def(imageLoader)
  },
  data() {
    return {
      status: status.initial
    }
  },
  beforeMount() {
    this.load()
  },
  beforeUpdated() {
    this.load()
  },
  methods: {
    ok() {
      this.status = status.transition(this.status, 'done').value
    },
    error() {
      this.status = status.transition(this.status, 'error').value
    },
    load() {
      this.__imageLoader__(this.url)
        .then(this.ok)
        .catch(this.error)
    }
  }
}
</script>
