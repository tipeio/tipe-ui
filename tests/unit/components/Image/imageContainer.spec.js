import { shallowMount } from '@vue/test-utils'
import ImageContainer from '@/components/image/ImageContainer.vue'

describe('Image Container', () => {
  it('name is TipeImageContainer', () => {
    const wrapper = shallowMount(ImageContainer)
    expect(wrapper.name()).toBe('TipeImageContainer')
  })

  it('image is loading', () => {
    const wrapper = shallowMount(ImageContainer)

    expect(wrapper.vm.status).toBe('waiting')
  })

  it('image has loaded', async () => {
    let res
    const Image = new Promise(resolve => {
      res = resolve
    })
    const imageLoader = () => Image
    const propsData = { imageLoader }
    const wrapper = shallowMount(ImageContainer, { propsData })

    res()
    await Image

    expect(wrapper.vm.status).toBe('ok')
  })

  it('image failed to load', async () => {
    let rej
    const Image = new Promise((resolve, reject) => {
      rej = reject
    })

    const imageLoader = () => Image
    const propsData = { imageLoader }
    const wrapper = shallowMount(ImageContainer, { propsData })

    try {
      await rej()
    } catch (error) {
      expect(wrapper.vm.status).toBe('error')
    }
  })
})
