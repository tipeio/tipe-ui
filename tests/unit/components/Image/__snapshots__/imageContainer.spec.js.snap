// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Container <template> matches previous snapshot 1`] = `<waitingtemplate-stub></waitingtemplate-stub>`;
